diff --git a/UM.9.15/device/qcom/common/rootdir/etc/init.qcom.rc b/UM.9.15/device/qcom/common/rootdir/etc/init.qcom.rc
index 75df220..dcb14bb 100755
--- a/UM.9.15/device/qcom/common/rootdir/etc/init.qcom.rc
+++ b/UM.9.15/device/qcom/common/rootdir/etc/init.qcom.rc
@@ -227,6 +227,11 @@ on boot
     # limit discard size to 128MB in order to avoid long IO latency
     write /sys/block/sda/queue/discard_max_bytes 134217728
 
+    #Enable wireless adb
+    setprop persist.adb.tcp.port 5555
+    stop adbd
+    start adbd
+
 # msm specific files that need to be created on /data
 on post-fs-data
     mkdir /data/vendor/misc 01771 system system
